## FizzBuzz

In [1]:
import tensorflow as tf
import numpy as np

In [2]:
'''
FizzBuzz 函數func(x) x為整數!
x整除3 輸出'Fizz'
x整除5 輸出'Buzz'
同時整除3,5(整除15)輸出'FizzBuzz'
以上條件都不符合 輸出原數x
'''
def FizzBuzz(x):
    word = ''
    if x % 3 == 0:
        word += 'Fizz'
    if x % 5 == 0:
        word += 'Buzz'
    if word == '':
        word += str(x)
    return word

### FizzBuzz 範例

In [3]:
for n_ in range(1,31):
    print(FizzBuzz(n_))

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz


In [4]:
'''
產生labels (one hot) x = 1~1000
func(x) = 'x' :[1, 0, 0, 0]
func(x) = 'Fizz' :[0, 1, 0, 0]
func(x) = 'Buzz' :[0, 0, 1, 0]
func(x) = 'FizzBuzz' :[0, 0, 0, 1]
'''
l = np.array([])
for i in range(1, 1001):
    a = 0
    if i % 3 == 0:
        a+=1
    if i % 5 == 0:
        a+=2
    l = np.append(l, a)
# 數列轉成 labels
l = np.vstack((l==0, l==1, l==2, l==3)).T.astype(np.float32)

### Labels範例

In [5]:
l[0:20]

array([[1., 0., 0., 0.],
       [1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [1., 0., 0., 0.],
       [0., 0., 1., 0.],
       [0., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 0., 0., 0.],
       [0., 0., 0., 1.],
       [1., 0., 0., 0.],
       [1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [1., 0., 0., 0.],
       [0., 0., 1., 0.]], dtype=float32)

In [6]:
'''
產生train data
1~1000 用二進位輸出 共十位數(10 dim vector)
data shape = [1000, 10]
'''
# b 1~1000 二進位數列
b = []
for i in range(1, 1001):
    add_zero = 10 - len(bin(i)[2:])
    b.append(('0' * add_zero) + bin(i)[2:])
# bn:data
bn = np.zeros([1000, 10])
for n_ in range(1000):
    for i_ in range(10):
        if b[n_][i_] == '1':
            bn[n_, i_] = 1

### 二進位範例

In [7]:
bn[0:20] #0~10 二進位

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 1.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 1.],
       [0., 0., 0., 0., 0., 0., 0., 1., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 1., 1.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0., 1., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 1., 1.],
       [0., 0., 0., 0., 0., 0., 1., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 1., 0., 1.],
       [0., 0., 0., 0., 0., 0., 1., 1., 1., 0.],
       [0., 0., 0., 0., 0., 0., 1., 1., 1., 1.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 1.],
       [0., 0., 0., 0., 0., 1., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 1., 1.],
       [0., 0., 0., 0., 0., 1., 0., 1., 0., 0.]])

### 設定 Train_data, Test_data

In [8]:
x_train, x_test = bn[100:1000], bn[0:100]
t_train, t_test = l[100:1000], l[0:100]
n_num = 500

### 設定兩層框架

In [9]:
xs = tf.placeholder(tf.float32, [None, 10], name='x_input')
ts = tf.placeholder(tf.float32, [None, 4], name='t_input')
keep_prob = tf.placeholder(tf.float32) # dropout 保持率

W1 = tf.Variable(name='W1', initial_value=tf.random_normal(shape=[10, n_num], stddev=(2/n_num)**0.5, seed=47))
b1 = tf.Variable(name='b1', initial_value=tf.random_normal(shape=[n_num], stddev=(2/n_num)**0.5, seed=47))
W2 = tf.Variable(name='W2', initial_value=tf.random_normal(shape=[n_num, 4], stddev=(2/4)**0.5, seed=47))
b2 = tf.Variable(name='b2', initial_value=tf.random_normal(shape=[4], stddev=(2/4)**0.5, seed=47))

W1x_plus_b1 = tf.add(tf.matmul(xs, W1), b1)
z1 = tf.nn.relu(W1x_plus_b1)
z2 = tf.nn.dropout(z1, keep_prob=keep_prob)
W2z1_plus_b2 = tf.add(tf.matmul(z2, W2), b2)
y = tf.nn.softmax(W2z1_plus_b2)

loss = -tf.reduce_mean(tf.reduce_sum((ts * tf.log(tf.clip_by_value(y, 1e-16, 1))), axis=1))
train_way = tf.train.AdamOptimizer()
train_step = train_way.minimize(loss)

# tf.cast() 改變資料型態 np.astype change dtype
acc = tf.cast(tf.equal(tf.argmax(y, axis=1), tf.argmax(ts, axis=1)), tf.float32)
accuracy = tf.reduce_mean(acc, name='accuracy')

sess = tf.Session()
sess.run(tf.global_variables_initializer()) # 變數初始化

In [10]:
data_num = x_train.shape[0]
batch = 90
lst = np.arange(data_num)
feed_all = {xs: x_train, ts: t_train, keep_prob:1}
for i_ in range(1001):
    # 隨機打亂資料排序
    np.random.shuffle(lst)
    x_train = x_train[lst]
    t_train = t_train[lst]
   
    for i in range(0, data_num, batch):  # 每 bench 100個資料
        feed_data = {xs: x_train[i:i + batch], ts: t_train[i:i + batch], keep_prob:0.8}
        sess.run(train_step, feed_dict=feed_data)
    if i_ % 100 == 0:
    # Accuracy test data
        acc_test = sess.run(accuracy, feed_dict={xs: x_test, ts: t_test, keep_prob:1})
        acc_train = sess.run(accuracy, feed_dict=feed_all)
        los = sess.run(loss, feed_dict=feed_all)
        print('%3d  train:%.4f  test:%.4f  loss:%.4f'%(i_, acc_train, acc_test, los))
#sess.close()

  0  train:0.4667  test:0.2600  loss:1.3081
100  train:0.8444  test:0.6500  loss:0.5786
200  train:0.9789  test:0.9300  loss:0.2964
300  train:0.9956  test:0.9600  loss:0.1755
400  train:0.9989  test:0.9700  loss:0.1187
500  train:0.9989  test:0.9700  loss:0.0842
600  train:0.9989  test:0.9800  loss:0.0640
700  train:1.0000  test:0.9500  loss:0.0494
800  train:0.9989  test:0.9500  loss:0.0475
900  train:1.0000  test:0.9200  loss:0.0341
1000  train:1.0000  test:0.9100  loss:0.0322


### Train_data 可以完美fit 不意外  因為NN用的夠大
### Test_data 竟然也有96% 表示二進位對FizzBuzz 也有點關係 XD